Enable I2C#6 instead of I2C#8

staroselskii opened the pull request:
Initial I2C#6 pins' configuration came from the firmware. Which quite likely has been used in the phones, where I2C#8, that is not part of Atom peripheral, is in use. Thus we need to override the leftover.

In order to achieve that I needed to port some of your Linux kernel code to U-Boot:

- scu_ipc_raw_command() that takes care of the complex command sent to MCU to let go off of I2C#6 pins
- add an adapted version of pinctrl-merrifield.c in pinmux.c that takes care of the above mentioned pins
- add an ACPI binding to make it usable by others

andy-shev commented:
Almost there, I would leave few comments, feel free to address them and resend this PR.

P.S. Looks pretty much cool! Thanks for your work!

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,100 @@
+	}
+}
+
+int mrfld_pinconfig_protected(unsigned int pin, u32 mask, u32 bits)
```
Since this is an API it has to get a pin control device first, something like

        ret = syscon_get_by_driver_data(X86_SYSCON_SCU, &dev);

The idea behind is to get lazy probe (that's called "Driver Model" in U-Boot).

staroselskii replied:
@andy-shev 
Thanks. Makes sense.

But it seems like probe is already automagically called just by the sheer addition of the probe callback, doesn't it? I'm pretty sure I'm missing something, of course.

andy-shev replied:
If you call function before device is probed, it wouldn't work and crash everything. ->probe() will be called when we try to get pointer to a device.

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,100 @@
+
+static int tangier_pinctrl_probe(struct udevice *dev)
+{
+    void *base_addr = syscon_get_first_range(X86_SYSCON_PINCONF);
```
Mind indentation.

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,100 @@
+
+int mrfld_pinconfig_protected(unsigned int pin, u32 mask, u32 bits)
+{
+	struct mrfld_family *mf = &i2c_family;
```
You need to keep your families somewhere else and here convert pin to family followed by family to bufcfg address.

staroselskii replied:
Okay. Something like `mrfld_get_family` from pinctrl-merrifield.c?

andy-shev replied:
Yes.

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,100 @@
+}
+
+static const struct udevice_id tangier_pinctrl_match[] = {
+	{ .compatible = "intel,pinctrl-tangier",
```
One line?
I mean compatible string with second parameter can be on one line. can they?

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,100 @@
+
+	ret = scu_ipc_raw_command(cmd, 0, &v, 4, NULL, 0, (u32) bufcfg, 0);
+	if (ret)
+		printf("Failed to set bufcfg via SCU IPC for pin %u (%d)\n", pin, ret);
```
Hmm... I don't remember if `err()` stops an execution, please check and use it instead.

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,100 @@
+
+	v = (value & ~mask) | (bits & mask);
+
+	debug("scu: cmd: %d v: 0x%x p: 0x%x bits: %d, mask: %d bufcfg: 0x%p\n",
```
I doubt it's useful in production. (Note ahead, I'm wondering if upstream will ask to use regmap for this driver, which, I guess, has a facility to dump registers)

staroselskii replied:
Not useful but this was extremely useful for me to get the parameters right. I'll get rid of it as soon as we get everything else.

andy-shev replied:
Fine to me as well, but should be removed before sending upstream.

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: GPL-2.0
```
Yep, like I said in the bug report, check _pinctrl-sandbox.c_ for a hint how this should look like. (Yes, I understand that we are creating a limited driver, it needs to be taken into account)

staroselskii replied:
Okay. So we're making it the UCLASS_PINCTRL, then? With some of pinctrl_ops implemented? And then we'll configure the pins in edison.c using the pinctrl API? Is this the plan? If so, I'll need some time adapting this code to the pinctrl API.

andy-shev replied:
I'm pretty sure that this what upstream ask us during review. However, we can try first with reduced solution. I guess, we may send it upstream and see what they will tell. In any case it would be nice to have your stuff in my tree sooner.

staroselskii commented:
@andy-shev 
Hello, Andy!

I addressed most of your concerns and made a force push. I also fixed most of the warnings checkpatch.pl gave along the way.

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,146 @@
+	unsigned int bufno;
+
+	family =  mrfld_get_family(pinctrl, pin);
+
```
Redundant blank line

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,146 @@
+};
+
+static const struct mrfld_family *mrfld_get_family(struct mrfld_pinctrl *mp,
+						   unsigned int pin)
```
I dunno if it's GH web or in your sources you have this indentation issue (on the second line `u` should be in the same column as `s` on the line above)

Another possibility is to split like

    static const struct mrfld_family *
    mrfld_get_fmaily(...)

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,146 @@
+}
+
+static void __iomem *mrfld_get_bufcfg(struct mrfld_pinctrl *pinctrl,
+			unsigned int pin)
```
Same about indentation

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,146 @@
+}
+
+static void mrfld_setup_families(void *base_addr, struct mrfld_family *families,
+			unsigned int nfam)
```
Same with indentation.

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,146 @@
+	v = (value & ~mask) | (bits & mask);
+
+	debug("scu: v: 0x%x p: 0x%x bits: %d, mask: %d bufcfg: 0x%p\n",
+			v, (u32)bufcfg, bits, mask, bufcfg);
```
Indentation.

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,146 @@
+			v, (u32)bufcfg, bits, mask, bufcfg);
+
+	ret = scu_ipc_raw_command(IPCMSG_INDIRECT_WRITE, 0, &v, 4,
+			NULL, 0, (u32)bufcfg, 0);
```
Ditto.

andy-shev commented on arch/x86/cpu/tangier/pinmux.c:
```diff
@@ -0,0 +1,146 @@
+	struct mrfld_pinctrl *pinctrl = dev_get_priv(dev);
+
+	mrfld_setup_families(base_addr, mrfld_families,
+			ARRAY_SIZE(mrfld_families));
```
One line?

andy-shev commented:
Please, address some minor issues and I'm ready to take it.

staroselskii commented:
@andy-shev 

Hi, I think I got it this time.

I set ts=8 sw=8 and see that I have scu_ipc_raw_command() properly aligned in Vim but GH shows them a little unaligned. Is this an issue?
